### PR TITLE
feat/emoji-with-instant-update

### PR DESCRIPTION
**Pull request by Hitesh-gurnani**

### CODE REVIEW

**Comment by Hitesh-gurnani** [comment]
https://github.com/townhall-gov/townhall/assets/55234408/9e1532b6-0dce-4df7-a3d3-354feb0b5ef3

**Comment by This-Is-Prince** [comment]
# Slack emoji design

Screenshot 2023-09-20 at 6 32 54 AM

# Townhall emoji design

Screenshot 2023-09-20 at 6 33 08 AM


why there is no space between `count` and `emoji`?

**Comment by This-Is-Prince** [comment]
Emoji reaction flow working randomly

[reaction.webm](https://github.com/townhall-gov/townhall/assets/89686019/0ff31b2e-fd5c-4bda-9e52-f1f9108d0606)

File: pages/api/posts/index.ts [hunk 1]
```diff
@@ -82,7 +82,11 @@ export const getPosts: TGetPostsFn = async (params) => {
 						const comments_count = data.comments_count || 0;
 						const reactions_count = {
 							[EReaction.DISLIKE]: data?.reactions_count?.[EReaction.DISLIKE] || 0,
-							[EReaction.LIKE]: data?.reactions_count?.[EReaction.LIKE] || 0
+							[EReaction.LIKE]: data?.reactions_count?.[EReaction.LIKE] || 0,
```
Comment (This-Is-Prince) [inline pages/api/posts/index.ts:85]: Please use syntax like
`const reactions_count = getDefaultReactions();`
`Object.values(EReaction).forEach((reaction) => {`
     `if (data.reactions_count?.[reaction]) {`
            `reactions_count[reaction] = data.reactions_count[reaction];`
     `}`
`});`

**Comment by This-Is-Prince** [comment]
Reacted by color

image

File: src/ui-components/EmojiTooltip.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	const { onClick } = props;
+	const formattedItemList = Object.values(EReaction).map((emoji) => (
+		<Button key={emoji} className='cursor-pointer p-0 bg-transparent border-none text-xl hover:bg-white' onClick={() => onClick(emoji)}>{emoji}</Button>
+	));
+	return (
+		<Tooltip
+			color='#66A5FF'
+			title={<div className='flex gap-2 bg-none'>{formattedItemList}</div>}
+		>
+			<span className='cursor-pointer'>
```
Comment (This-Is-Prince) [inline src/ui-components/EmojiTooltip.tsx:22]: Is there any specific reason to use extra `span`?

File: src/ui-components/EmojiTooltip.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import { EReaction } from '~src/types/enums';
+import { AddEmojiReactionIcon } from './CustomIcons';
+
+interface IEmojiTooltipProps {
+	onClick: (type: EReaction) => Promise<void>;
+}
+const EmojiTooltip:FC<IEmojiTooltipProps> = (props) => {
+	const { onClick } = props;
+	const formattedItemList = Object.values(EReaction).map((emoji) => (
+		<Button key={emoji} className='cursor-pointer p-0 bg-transparent border-none text-xl hover:bg-white' onClick={() => onClick(emoji)}>{emoji}</Button>
```
Comment (This-Is-Prince) [inline src/ui-components/EmojiTooltip.tsx:15]: code is not formatted correctly
<Button
key
className
onClick
/>

File: src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx [hunk 4]
```diff
@@ -74,23 +93,55 @@ const Reactions = () => {
 	}
 	return (
 		<div className='flex items-center gap-x-[15px]'>
-			<Reaction
+			<EmojiCounts
```
Comment (This-Is-Prince) [inline src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx:96]: Please use `for loop` for rendering the multiple EmojiCounts component

File: src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx [hunk 5]
```diff
@@ -25,18 +26,36 @@ const Reactions = () => {
...
 			connectWallet();
 			return;
 		}
 		if (!isRoomFollow) {
 			joinRoom();
 			return;
 		}
+		let isDeleted=false;
+		post?.reactions.map((reaction) => {
+			if (reaction.type==type && reaction.user.id==user.id) {
```
Comment (This-Is-Prince) [inline src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx:39]: Please use user `address`

File: src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx [hunk 6]
```diff
@@ -48,7 +67,7 @@ const Reactions = () => {
 					const error = 'Something went wrong, unable to give a reaction to post.';
```
Comment (This-Is-Prince) [inline src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx:67]: If error comes why you are not removing reaction from reactions array.

File: src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx [hunk 7]
```diff
@@ -48,7 +67,7 @@ const Reactions = () => {
...
 					dispatch(postActions.setError(error));
 				} else {
-					dispatch(postActions.setReaction(data));
+					// dispatch(postActions.setReaction(data));
```
Comment (This-Is-Prince) [inline src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx:70]: Please set valid document again

File: src/ui-components/EmojiCounts.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    type: EReaction;
+    userReaction?:EReaction | null,
+    reactions:IReactionWithUser[]
+}
+const EmojiCounts:FC<IEmojiEmojiCountsProps> = (props) => {
+	const { onClick , count , loading , type , reactions  } = props;
+	const { user } = useProfileSelector();
+	const findUser = reactions.find((reaction) => (user?.address == reaction.user.address && type == reaction.type));
+	const content = (
+		<div style={{  maxHeight: '100px', minWidth:'200px', overflowY: 'auto' }}>
```
Comment (This-Is-Prince) [inline src/ui-components/EmojiCounts.tsx:25]: I think we can use tailwindcss classes here.
This div is not required

File: src/ui-components/EmojiCounts.tsx [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+												ethIdenticonSize: 20,
+												identiconSize: 20
+											}}
+										/>
+									</li>
+								);
+							})
+						}
+					</ul>
+					: <p>No reactions found</p>
```
Comment (This-Is-Prince) [inline src/ui-components/EmojiCounts.tsx:49]: If there is no reactions why you are rendering this component (Very Bad practice)

File: src/ui-components/EmojiCounts.tsx [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+			}
+		</div>
+	);
+	return (
+		<Popover content={content} title={<div className='text-[#0E2D59] text-lg'>Reacted By</div>}>
+			<button
+				onClick={() => onClick(type)}
+				disabled={loading}
+				className={
+					classNames(`${count<=0?'hidden':''} flex text-white border border-solid border-blue_primary rounded-2xl py-0.5 hover:bg-blue_primary px-2 font-medium text-sm  ${findUser?'bg-blue_primary':'bg-transparent'} `, {
```
Comment (This-Is-Prince) [inline src/ui-components/EmojiCounts.tsx:59]: Don't use hidden, instead if reactions is not found don't render this component

File: src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/RepliesWrapper/Replies/Reply/Footer/Reactions/index.tsx [hunk 11]
```diff
@@ -91,24 +110,55 @@ const ReplyReactions: FC<IReplyReactionsProps> = (props) => {
...
+			<EmojiCounts
 				count={dislikeReactions?.length || 0}
 				loading={loading}
-				onReaction={onReaction}
+				onClick={onReaction}
 				type={EReaction.DISLIKE}
 				userReaction={userReaction}
 				reactions={dislikeReactions}
 			/>
+			<EmojiCounts
```
Comment (This-Is-Prince) [inline src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/RepliesWrapper/Replies/Reply/Footer/Reactions/index.tsx:129]: Please check post files comment

File: src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/Footer/Reactions/index.tsx [hunk 12]
```diff
@@ -83,24 +101,55 @@ const CommentReactions: FC<ICommentReactionsProps> = (props) => {
...
+			<EmojiCounts
 				count={dislikeReactions?.length || 0}
 				loading={loading}
-				onReaction={onReaction}
+				onClick={onReaction}
 				type={EReaction.DISLIKE}
 				userReaction={userReaction}
 				reactions={dislikeReactions}
 			/>
+			<EmojiCounts
```
Comment (This-Is-Prince) [inline src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/Footer/Reactions/index.tsx:120]: Please check post files comment

File: src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/Footer/Reactions/index.tsx [hunk 13]
```diff
@@ -28,23 +31,43 @@ const CommentReactions: FC<ICommentReactionsProps> = (props) => {
...
-		if (!isLoggedIn) {
+		if (!isLoggedIn || !user) {
 			connectWallet();
 			return;
 		}
 		if (!isRoomFollow) {
 			joinRoom();
 			return;
 		}
+		let isDeleted=false;
```
Comment (This-Is-Prince) [inline src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/Footer/Reactions/index.tsx:46]: Can you please create a custom function for this, you are using same code 3 times.

**Review by This-Is-Prince — CHANGES_REQUESTED** [review]
Please test properly, now reaction feature completely broke

**Comment by Hitesh-gurnani** [comment]
> Emoji reaction flow working randomly
> 
>  reaction.webm

This itself is the required output

File: src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/Footer/Reactions/index.tsx (hunk 13 above)
Comment (Hitesh-gurnani) [reply src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/Footer/Reactions/index.tsx:46]:
```
const getIsDeletedField = (reactions: IReactionWithUser[],type:EReaction,user:IUserInfo) => {
	let isDisabled:boolean=false;
	reactions.map((reaction) => {
		if (reaction.type==type && reaction.user.address==user.address) {
			isDisabled=true;
		}
	},[]);
	return isDisabled;
};
```

File: src/ui-components/EmojiTooltip.tsx (hunk 2 above)
Comment (Hitesh-gurnani) [reply src/ui-components/EmojiTooltip.tsx:22]: `<AddEmojiReactionIcon className='text-xl cursor-pointer text-blue_primary'/>`

File: src/ui-components/EmojiTooltip.tsx (hunk 3 above)
Comment (Hitesh-gurnani) [reply src/ui-components/EmojiTooltip.tsx:15]:
```
const formattedItemList = Object.values(EReaction).map((emoji) => (
		<Button
			key={emoji}
			className='cursor-pointer p-0 bg-transparent border-none text-xl hover:bg-white'
			onClick={() => onClick(emoji)}
		>
			{emoji}
		</Button>
	));
```

File: src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx (hunk 5 above)
Comment (Hitesh-gurnani) [reply src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx:39]:
```
const getIsDeletedField = (reactions: IReactionWithUser[],type:EReaction,user:IUserInfo) => {
	let isDisabled:boolean=false;
	reactions.map((reaction) => {
		if (reaction.type==type && reaction.user.address==user.address) {
			isDisabled=true;
		}
	},[]);
	return isDisabled;
};
```

File: src/ui-components/EmojiCounts.tsx (hunk 8 above)
Comment (Hitesh-gurnani) [reply src/ui-components/EmojiCounts.tsx:25]: `<div className='min-h-[100px] min-w-[200px] overflow-y-auto'>`

File: src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/RepliesWrapper/Replies/Reply/Footer/Reactions/index.tsx (hunk 11 above)
Comment (Hitesh-gurnani) [reply src/components/Room/Post/ContentWrapper/CommentsWrapper/Comments/Comment/RepliesWrapper/Replies/Reply/Footer/Reactions/index.tsx:129]:
```
const reactionsTypes = useCommentReactions(reactions);

{
				reactionsTypes.map((reaction,idx) => {
					return (
						<EmojiCounts
							key={idx}
							count={reaction.reactionType.length || 0}
							loading={loading}
							onClick={onReaction}
							type={reaction.type}
							userReaction={userReaction}
							reactions={reaction.reactionType}
						/>
					);
				})
			}
```

```
const useCommentReactions = (reactions: IReactionWithUser[]) => {
	const clapReactions: IReactionWithUser[] = [];
	const dislikeReactions: IReactionWithUser[] = [];
	const happyReactions: IReactionWithUser[] = [];
	const heartReactions: IReactionWithUser[] = [];
	const likeReactions: IReactionWithUser[] = [];
	const powerReactions: IReactionWithUser[] = [];
	if (reactions && Array.isArray(reactions)) {
		reactions.forEach((reaction) => {
			if (reaction.type === EReaction.CLAP) {
				clapReactions.push(reaction);
			}
			else if (reaction.type === EReaction.DISLIKE) {
				dislikeReactions.push(reaction);
			}
			else if (reaction.type === EReaction.HAPPY) {
				happyReactions.push(reaction);
			}
			else if (reaction.type === EReaction.HEART) {
				heartReactions.push(reaction);
			}
			else if (reaction.type === EReaction.LIKE) {
				likeReactions.push(reaction);
			}
			else if (reaction.type === EReaction.POWER) {
				powerReactions.push(reaction);
			}

		});
	}
	return [{ reactionType:clapReactions,type:EReaction.CLAP } , { reactionType:dislikeReactions,type:EReaction.DISLIKE } , { reactionType:happyReactions,type:EReaction.HAPPY } , { reactionType:heartReactions,type:EReaction.HEART } , { reactionType:likeReactions,type:EReaction.LIKE } , { reactionType:powerReactions,type:EReaction.POWER }];
};
```

File: src/ui-components/EmojiCounts.tsx (hunk 9 above)
Comment (Hitesh-gurnani) [reply src/ui-components/EmojiCounts.tsx:49]:
```
const content = (
		<div className="min-h-[100px] min-w-[200px] overflow-y-auto">
			{reactions && reactions.length > 0 && (
				<ul className="flex flex-col w-full justify-center gap-y-3 mt-3">
					{reactions.map((reaction) => (
						<li key={reaction.id} className="list-none">
							<UserInfo
								user_info={reaction.user}
								addressProps={{
									className: 'text-grey_tertiary font-medium text-base leading-[20px] tracking-[0.01em]',
									ethIdenticonSize: 20,
									identiconSize: 20
								}}
							/>
						</li>
					))}
				</ul>
			)}
		</div>
	);
```

File: src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx (hunk 4 above)
Comment (Hitesh-gurnani) [reply src/components/Room/Post/ContentWrapper/Content/Reactions/index.tsx:96]:
```
{
				reactionsTypes.map((reaction,idx) => {
					return (
						<EmojiCounts
							key={idx}
							count={reaction.reactionType.length || 0}
							loading={loading}
							onClick={onReaction}
							type={reaction.type}
							userReaction={userReaction}
							reactions={reaction.reactionType}
						/>
					);
				})
			}
```

File: src/ui-components/EmojiCounts.tsx (hunk 10 above)
Comment (Hitesh-gurnani) [reply src/ui-components/EmojiCounts.tsx:59]:
```
{
				reactionsTypes.map((reaction, idx) => (
					reaction.reactionType.length > 0 && (
						<EmojiCounts
							key={idx}
							count={reaction.reactionType.length || 0}
							loading={loading}
							onClick={onReaction}
							type={reaction.type}
							userReaction={userReaction}
							reactions={reaction.reactionType}
						/>
					)
				))
			}
```

**Pull request closed without merging**